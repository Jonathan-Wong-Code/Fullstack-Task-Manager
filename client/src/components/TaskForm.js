import React, { useReducer } from "react";
import { withRouter } from "react-router-dom";

import { useTaskDispatch, useTaskState } from "./../context/task-context";
import { createTask, editTask } from "./../async-helpers/tasks";

const reducer = (state, newState) => {
  return { ...state, ...newState };
};

function TaskForm({ type, history, editedTask }) {
  const [{ title, description, completed, _id }, setState] = useReducer(
    reducer,
    {
      description: editedTask ? editedTask.description : "",
      completed: editedTask ? editedTask.completed : false,
      title: editedTask ? editedTask.title : "",
      _id: editedTask ? editedTask._id : undefined
    }
  );

  const taskDispatch = useTaskDispatch();
  const { error } = useTaskState();

  const handleSubmit = async e => {
    e.preventDefault();
    const task = {
      title,
      description,
      completed,
      _id
    };

    let success;
    if (type === "create") {
      success = await createTask(task, taskDispatch);
    } else {
      success = await editTask(task, taskDispatch);
    }
    if (success) {
      history.push("/dashboard");
    }
  };

  return (
    <>
      <form onSubmit={handleSubmit}>
        <label htmlFor="title">Title:</label>
        <input
          type="text"
          id="title"
          value={title}
          onChange={e => setState({ title: e.target.value })}
        />
        <label htmlFor="description">Description:</label>
        <textarea
          type="text"
          id="description"
          rows="10"
          cols="50"
          value={description}
          onChange={e => setState({ description: e.target.value })}
        />
        <label htmlFor="completed">completed?</label>
        <input
          id="completed"
          type="checkbox"
          checked={completed}
          value={completed}
          onChange={e => setState({ completed: e.target.value })}
        />
        <button> {type === "create" ? "Create" : "Edit"}</button>
      </form>
      {error && <p>{error}</p>}
    </>
  );
}

export default withRouter(TaskForm);
