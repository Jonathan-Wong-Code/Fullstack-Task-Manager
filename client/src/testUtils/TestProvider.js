import React from "react";
import { UserProvider } from "./../context/user-context";
import { AuthProvider } from "./../context/auth-context";
import { TaskProvider } from "./../context/task-context";

function TestProvider({ children }) {
  return (
    <AuthProvider>
      <UserProvider>
        <TaskProvider>{children}</TaskProvider>
      </UserProvider>
    </AuthProvider>
  );
}

export default TestProvider;
