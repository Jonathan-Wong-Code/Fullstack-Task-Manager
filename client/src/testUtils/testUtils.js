import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import { render } from "@testing-library/react";
import { UserProvider } from "./../context/user-context";
import { AuthProvider } from "./../context/auth-context";
import { TaskProvider } from "./../context/task-context";
import React from "react";
export const renderRouter = (
  ui,
  {
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] }),
    ...options
  }
) => {
  return {
    ...render(<Router history={history}>{ui}</Router>, options),
    history
  };
};

export function renderWithContextRouter(
  ui,
  {
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] }),
    authValue,
    userValue,
    taskValue,
    ...options
  }
) {
  return {
    ...render(
      <AuthProvider value={authValue}>
        <UserProvider value={userValue}>
          <TaskProvider value={taskValue}>
            <Router history={history}>{ui}</Router>
          </TaskProvider>
        </UserProvider>
      </AuthProvider>,
      options
    ),
    history
  };
}
