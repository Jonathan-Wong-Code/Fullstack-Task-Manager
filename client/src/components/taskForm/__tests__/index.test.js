import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { renderContext } from "./../../../testUtils/testUtils";
import TaskForm from "..";
import axios from "axios";
import { fireEvent, wait, cleanup } from "@testing-library/react";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";

jest.mock("axios");

const history = createMemoryHistory();

const mockTask = {
  title: "Test task",
  description: "Test Descript",
  completed: false,
  _id: "test1234"
};

const response = {
  data: {
    data: {
      task: mockTask
    }
  }
};
afterEach(cleanup);

describe("<TaskForm>", () => {
  beforeEach(() => {
    history.push = jest.fn();
  });

  test("it should successfully create a new task", async () => {
    axios.mockImplementation(() => Promise.resolve(response));
    const { getByLabelText, getByTestId, queryByTestId } = renderContext(
      <Router history={history}>
        <TaskForm history={history} type="create" />,
      </Router>,
      {
        route: "/create"
      }
    );

    const title = getByLabelText(/title:/i);
    const description = getByLabelText(/description:/i);
    const completed = getByLabelText("completed?");

    fireEvent.change(title, { target: { value: mockTask.title } });
    fireEvent.change(description, { target: { value: mockTask.description } });
    fireEvent.click(completed);

    expect(title).toHaveValue(mockTask.title);
    expect(description).toHaveValue(mockTask.description);
    expect(completed).toHaveAttribute("value", "true");

    fireEvent.submit(getByTestId("task-form"));
    await wait(() => {
      expect(queryByTestId("task-form-error")).toBeNull();
      expect(history.push).toHaveBeenCalledTimes(1);
    });
  });

  test("it should throw an error if create task fails", async () => {
    const error = new Error();
    error.response = { data: { message: "failure" } };
    axios.mockImplementationOnce(() => Promise.reject(error));

    const { getByLabelText, getByTestId, container } = renderContext(
      <Router history={history}>
        <TaskForm history={history} type="create" />
      </Router>,
      {}
    );

    const title = getByLabelText(/title:/i);
    const description = getByLabelText(/description:/i);
    const completed = getByLabelText("completed?");

    fireEvent.change(title, { target: { value: mockTask.title } });
    fireEvent.change(description, {
      target: { value: mockTask.description }
    });
    fireEvent.click(completed);
    fireEvent.submit(getByTestId("task-form"));

    await wait(() => {
      expect(getByTestId("task-form-error")).toBeTruthy();
      expect(getByTestId("task-form-error").textContent).toBe("failure");
    });

    expect(history.push).not.toHaveBeenCalled();
  });
});
