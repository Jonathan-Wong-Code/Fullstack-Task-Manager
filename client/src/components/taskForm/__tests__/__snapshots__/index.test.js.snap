// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskForm> it should throw an error if create task fails 1`] = `
<div>
  <form
    data-testid="task-form"
  >
    <label
      for="title"
    >
      Title:
    </label>
    <input
      id="title"
      type="text"
      value="Test task"
    />
    <label
      for="description"
    >
      Description:
    </label>
    <textarea
      cols="50"
      id="description"
      rows="10"
      type="text"
    >
      Test Descript
    </textarea>
    <label
      for="completed"
    >
      completed?
    </label>
    <input
      id="completed"
      type="checkbox"
      value="true"
    />
    <button>
       
      Create
    </button>
  </form>
  <p
    data-testid="task-form-error"
  >
    failure
  </p>
</div>
`;
